### PR TITLE
Rewrite the update script in Python

### PR DESCRIPTION
**Pull request by alexcrichton**
With various updates to this script such as having a different upstream for `memory64` and otherwise including the `wasm-3.0` test suite this script was getting relatively complicated which is typically not a great situation for any bash script to be in. Instead rewrite this script in Python to avoid many of bash's pitfalls and be a bit more readable/reviewable/modifiable.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Inspired by discussion on https://github.com/WebAssembly/testsuite/pull/110

File: update-testsuite.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    repos.append(Repo('gc'))
+    repos.append(Repo('tail-call'))
+    repos.append(Repo('annotations'))
+    repos.append(Repo('function-references'))
+    repos.append(Repo('memory64', upstream=spec3))
+    repos.append(Repo('extended-const'))
+    repos.append(Repo('multi-memory'))
+    repos.append(Repo('relaxed-simd'))
+    repos.append(Repo('custom-page-sizes'))
+    repos.append(Repo('wide-arithmetic'))
```
Comment (sbc100) [inline update-testsuite.py:31]: Can you just declare this as single list:

````
repos = [
   spec3,
   Repo('threads'),
   ...
]
```

File: update-testsuite.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    # Make sure that `repos/spec` is a git repository
+    if not os.path.isdir('repos/spec'):
+        git('init', 'repos/spec')
```
Comment (sbc100) [inline update-testsuite.py:35]: If these is just one repo these days maybe we don't need two levels of nesting here?

File: update-testsuite.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        self.args = args
+
+    def __str__(self):
+        desc = f'failed to run: git {" ".join(self.args)}'
+        desc += f"\n\treturncode: {self.output.returncode}"
+        if len(self.output.stdout) > 0:
+            stdout = self.output.stdout.decode('utf-8')
+            stdout = stdout.replace('\n', '\n\t\t')
+            desc += f"\n\tstdout:\n\t\t{stdout}"
+        return desc
```
Comment (sbc100) [inline update-testsuite.py:199]: I kind of prefer to have "things used by X" declared earlier in file to "X" itself.    So maybe we could have `GitError` come first, then `get` etc?   Unless you feel the opposite?

File: update-testsuite.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                    tests.append((path, repo_path, dst))
+        return tests
+
+
+# Helper to run a `git` command and handle the output/exit code in an ergonomic
+# fashion.
+def git(*args, quiet=False):
+    if not quiet:
+        print('running: git', *args)
+    ret = subprocess.run(["git", *args], stdout=subprocess.PIPE)
```
Comment (sbc100) [inline update-testsuite.py:181]: There seems to be a mix of single and double quotes here.  I don't feel strongly which why but I prefer consistency.

File: update-testsuite.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    if ret.returncode != 0:
+        raise GitError(ret, args)
+    return ret.stdout.decode('utf-8').strip()
```
Comment (sbc100) [inline update-testsuite.py:184]: I think you can add `text=True` to the `.run` call above and skip this decode step.

File: update-testsuite.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    if len(updated) == 0:
+        return
+    message = "Update repos:\n\n"
+    for repo in updated:
+        message += f"  {repo.dir}:\n"
+        message += f"    {repo.url()}/commit/{repo.rev}\n"
+    message += "\n"
+    message += "This change was automatically generated by `update-testsuite.py`"
+    git('commit', '-a', '-m', message)
+
```
Comment (sbc100) [inline update-testsuite.py:95]: Python style (at least flake8) is to have two empty lines between top level functions/classes.

File: update-testsuite.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        git('add', dstdir)
+        status = git('status', '-s', dstdir)
+        if len(status) > 0:
+            updated.append(repo)
+
+    for repo in failed_merges:
+        print('!! failed to update:', repo.url())
+
+    # If anything was updated, make a commit message indicating as such.
+    if len(updated) == 0:
```
Comment (sbc100) [inline update-testsuite.py:86]: Should we print something here?  e.g. "No spec changes found, not creating a new commit"?

File: update-testsuite.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        desc += f"\n\treturncode: {self.output.returncode}"
+        if len(self.output.stdout) > 0:
+            stdout = self.output.stdout.decode('utf-8')
+            stdout = stdout.replace('\n', '\n\t\t')
+            desc += f"\n\tstdout:\n\t\t{stdout}"
+        return desc
+
+
+if __name__ == '__main__':
+    main()
```
Comment (sbc100) [inline update-testsuite.py:203]: I normally do `sys.exit(main())` here and then return 0 (or non-zero) from main on failure.

File: update-testsuite.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+
+# Script to automatically update all tests in this repository based on the
+# status of the upstream repositories themselves. This will prepare a git
+# submodule-of-sorts (not literally) in `repos/spec` with all the upstream repos
+# fetched into that one location. Tests will be diff'd against the merged
+# version of the upstream spec repo and the proposal repo, and if there's a
+# difference then the test is included.
```
Comment (sbc100) [inline update-testsuite.py:8]: Put this top level comment in a multi-lint doc string?

```
"""Script to automatically update all tests in 
...
"""
```

File: update-testsuite.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+                        continue
+                    _, dst = os.path.split(path)
+                    if subdir == 'legacy':
+                        dst = 'legacy/' + dst
+                    tests.append((path, repo_path, dst))
+        return tests
+
+
+# Helper to run a `git` command and handle the output/exit code in an ergonomic
+# fashion.
```
Comment (sbc100) [inline update-testsuite.py:177]: Move into a docstring inside the function?

File: update-testsuite.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+    def __str__(self):
+        repo = f'WebAssembly/{self.repo}'
+        if self.branch != 'main':
+            repo += f'@{self.branch}'
+        return repo
+
+    # Fetch the latest revision from this repository and store the revision
+    # within `self` of what was found.
+    def update(self):
```
Comment (sbc100) [inline update-testsuite.py:114]: Move comment into docstring here and elsewhere this class?

File: update-testsuite.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        self.git('fetch', self.url(), self.branch)
+        self.rev = self.git('rev-parse', 'FETCH_HEAD', quiet=True)
+
+    def url(self):
+        return f'https://github.com/WebAssembly/{self.repo}'
+
+    # Helper to run a `git` command within the `repos/spec` repository.
+    def git(self, *args, **kwargs):
+        return git('-C', 'repos/spec', *args, **kwargs)
```
Comment (sbc100) [inline update-testsuite.py:123]: This seems unrelated to any specific repo so maybe move it outside the class or at least mark it as `@staticmethod`

File: update-testsuite.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+        tests = []
+        for subdir in ['core', 'legacy']:
+            for root, dirs, files in os.walk(f'repos/spec/test/{subdir}'):
+                for file in files:
+                    path = os.path.join(root, file)
+                    repo_path = os.path.relpath(path, 'repos/spec')
+                    _, ext = os.path.splitext(path)
+                    if ext != '.wast':
+                        continue
+                    _, dst = os.path.split(path)
```
Comment (sbc100) [inline update-testsuite.py:169]: ```dst = os.path.basename(path)```

File: update-testsuite.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    #
+    # This is used to run diffs and copy the test to its final location.
+    def list_tests(self):
+        tests = []
+        for subdir in ['core', 'legacy']:
+            for root, dirs, files in os.walk(f'repos/spec/test/{subdir}'):
+                for file in files:
+                    path = os.path.join(root, file)
+                    repo_path = os.path.relpath(path, 'repos/spec')
+                    _, ext = os.path.splitext(path)
```
Comment (sbc100) [inline update-testsuite.py:166]: Its basically the same by maybe ```ext = os.path.splitext(path)[1]``` is slightly more readable?

**Review by sbc100 — COMMENTED** [review]
Thanks for working on this!

lgtm % a whole bunch of python nits.

File: update-testsuite.py (hunk 3 above)
Comment (alexcrichton) [reply update-testsuite.py:199]: Personally I prefer the style of "high level first" followed by more detailed later (e.g. the inverse of C), but happy to switch things around too, it's not a strong preference

**Comment by alexcrichton** [comment]
Thanks for the review! I also fixed an issue where if the spec repo itself had changed tests a commit wasn't made, but that should be fixed now too. (also apologies for the "clearly I'm not a Python programmer" python)

**Review by sbc100 — APPROVED** [review]
lgtm

@sunfishcode are you OK with this change?

**Review by tlively — COMMENTED** [review]
LGTM, thanks for doing this 🎉

**Comment by sunfishcode** [comment]
I haven't reviewed the code myself yet, but I have no special concerns here, and I trust the judgement of the people who have already reviewed it.

**Comment by alexcrichton** [comment]
Ok I'm going to go ahead and merge this and I'll be on the hook for any failures that happen in the near future from this refactoring...

**Comment by alexcrichton** [comment]
Oh wow it's significantly faster than before! The [last auto-update run](https://github.com/WebAssembly/testsuite/actions/runs/11940915581) was 5.5 minutes and a [manually triggered run](https://github.com/WebAssembly/testsuite/actions/runs/11962342574/job/33350534105) was 0.5 minutes.

Looks like stdout/stderr are interleaved there which isn't great though, but otherwise it seems to have at least run successfully in CI which is nice.
